frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,367 @@
+      })
+      break
+    }
+  }
+}
+
+sbp('chelonia/defineContract', {
+  name: 'gi.contracts/chatroom',
+  metadata: {
+    validate: objectOf({
+      createdDate: string, // action created date
+      username: string, // action creator
+      identityContractID: string // action creator identityContractID
+    }),
+    create () {
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: chatRoomType,
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
+          settings: {
+            messagesPerPage: CHATROOM_MESSAGES_PER_PAGE,
+            maxNameLength: CHATROOM_NAME_LIMITS_IN_CHARS,
+            maxDescriptionLength: CHATROOM_DESCRIPTION_LIMITS_IN_CHARS
+          },
+          attributes: {
+            creator: meta.username,
+            deletedDate: null,
+            archievedDate: null
+          },
+          users: {},
+          messages: []
+        }, data)
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
+        }
+      },
+      sideEffect ({ contractID, meta }) {
+        const rootState = sbp('state/vuex/state')
+        // this condition makes users redirect to the newly-created channel only after they created a new channel
+        if (meta.username === rootState.loggedIn.username) {
+          sbp('okTurtles.events/once', CONTRACT_IS_SYNCING, (cID, isSyncing) => {
+            if (cID === contractID && !isSyncing) {
+              sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
+            }
+          })
+        }
+      }
+    },
+    'gi.contracts/chatroom/join': {
+      validate: objectOf({
+        username: string // username of joining member
+      }),
+      process ({ data, meta, hash }, { state, getters }) {
+        const { username } = data
+        if (state.users[username] && !sbp('okTurtles.data/get', 'JOINING_CHATROOM')) {
+          throw new Error('Can not join the chatroom which you are already part of')
+        }
```

Comment:
> `!sbp('okTurtles.data/get', 'JOINING_CHATROOM')`


... I don't think this type of check should ever happen in a process function. AFAIK this should only ever appear in sideEffects.

Reply:
I think it works even inside the `process` function because `sbp` itself is what we import at the top, not a parameter sent to the `process` function.
Also I have tested if `sbp` works in `process` function on my side and confirmed it works.

Reply:
Finally, I understood your point and fixed it. :grin: